### PR TITLE
enable setting attributes when building an assoc

### LINKED ISSUE #532: How about adding attributes to `build` function?
So that we can use the built struct more easily.

### PR DESCRIPTION
**Pull request by tony612**
resolves #532

### CODE REVIEW

**Comment by paulcsmith** [comment]
This will be really helpful! Thanks @tony612

File: lib/ecto/association.ex [hunk 1]
```diff
@@ -295,9 +295,10 @@ defmodule Ecto.Association.Has do
 
   @doc false
   def build(%{assoc: assoc, owner_key: owner_key, assoc_key: assoc_key,
-              queryable: queryable}, struct) do
+              queryable: queryable}, struct, attributes \\ %{}) do
```
Comment (josevalim) [inline lib/ecto/association.ex]: We don't need to have the default argument in here because this is always invoked as a callback.

Comment (josevalim) [inline lib/ecto/association.ex]: Same for the builds below. :)

File: lib/ecto/association.ex [hunk 2]
```diff
@@ -295,9 +295,10 @@ defmodule Ecto.Association.Has do
...
     assoc
     |> apply(:__struct__, [])
+    |> Map.merge(Enum.into(attributes, %{}))
```
Comment (josevalim) [inline lib/ecto/association.ex]: `struct/2` does exactly this! This:

     assoc
     |> apply(:__struct__, [])
     |> Map.merge(Enum.into(attributes, %{}))

Is equivalent to:

     assoc
     |> struct(attributes)

:)

File: lib/ecto/model.ex [hunk 3]
```diff
@@ -96,10 +96,22 @@ defmodule Ecto.Model do
...
+      %Comment{id: nil, post_id: 13, text: "cool"}
+
+      iex> build(post, :comments, %{text: "cool"})
+      %Comment{id: nil, post_id: 13, text: "cool"}
+
+      iex> build(post, :comments, post_id: 1)
+      %Comment{id: nil, post_id: 13}
   """
-  def build(%{__struct__: model} = struct, assoc) do
+  def build(%{__struct__: model} = struct, assoc, attributes \\ %{}) do
```
Comment (josevalim) [inline lib/ecto/model.ex]: Default arguments here are great. :+1:

File: lib/ecto/model.ex [hunk 4]
```diff
@@ -96,10 +96,22 @@ defmodule Ecto.Model do
       %Comment{id: 13, post_id: 25}
       iex> build(comment, :post)
       %Post{id: nil}
+
+  You can also pass the attributes, which can be a map or
+  a keyword list, to set the struct's fields except the key.
```
Comment (josevalim) [inline lib/ecto/model.ex]: "except the association key" would be clearer. :)

**Comment by josevalim** [comment]
I have added some minor comments, thank you! <3

**Comment by tony612** [comment]
@josevalim Done. Please review. And `struct` function is very neat!
@paulcsmith Thanks! :smile:

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Pull request merged**